### PR TITLE
[LegalizeTypes][RISCV][X86] Legalize FP_ROUND to libcall in SoftPromoteHalfRes_FP_ROUND if the input type is softened.

### LINKED ISSUE #107607: [RISCV] Miscompilation on rv64
The following program miscompiles when using clang++ for rv64:
```
$ cat foo.cpp 
#include <cmath>

int main() {
  long double y = NAN;
  _Float16 a = static_cast<_Float16>(y);
  return (a != a);
}
```
The cast of `NAN` from `long double` to `_Float16` generates `0`, instead of `NAN`.

It only happens with clang++ in rv64. It works just fine in intel 64 and rv32. Compiling with g++ for rv64 also works.

### PR DESCRIPTION
**Pull request by topperc**
Previously we created an FP_TO_FP16 and legalized it in SoftenFloatOp_FP_ROUND. This caused i16 to be sent to call lowering instead of f16. This results in the ABI not being followed if f16 is supposed to be passed in a different register than i16.

Looking at the libgcc code for the library function it appears the value is returned in xmm0 so the X86 test was being miscompiled before.

Fixes #107607.

### CODE REVIEW

File: llvm/test/CodeGen/X86/bfloat.ll [hunk 1]
```diff
@@ -1983,8 +1983,6 @@ define bfloat @PR115710(fp128 %0) nounwind {
 ; X86-NEXT:    vmovaps {{[0-9]+}}(%esp), %xmm0
 ; X86-NEXT:    vmovups %xmm0, (%esp)
 ; X86-NEXT:    calll __trunctfbf2
-; X86-NEXT:    # kill: def $ax killed $ax def $eax
-; X86-NEXT:    vmovw %eax, %xmm0
```
Comment (phoebewang) [inline llvm/test/CodeGen/X86/bfloat.ll:1987]: Thanks! I should have noticed the problem during the review. The interesting thing is the `fp80 -> bf16` is correct..

**Review by phoebewang — APPROVED** [review]
LGTM, though it's a bit confusing to see a BF16 test affected by `SoftPromoteHalfRes_FP_ROUND`.

File: llvm/test/CodeGen/RISCV/half-convert.ll [hunk 2]
```diff
@@ -8587,3 +8587,811 @@ start:
...
+; CHECK64-IZDINXZHINXMIN-NEXT:    call __extendsftf2
+; CHECK64-IZDINXZHINXMIN-NEXT:    ld ra, 8(sp) # 8-byte Folded Reload
+; CHECK64-IZDINXZHINXMIN-NEXT:    .cfi_restore ra
+; CHECK64-IZDINXZHINXMIN-NEXT:    addi sp, sp, 16
+; CHECK64-IZDINXZHINXMIN-NEXT:    .cfi_def_cfa_offset 0
+; CHECK64-IZDINXZHINXMIN-NEXT:    ret
+  %a = load half, ptr %x
+  %b = fpext half %a to fp128
+  ret fp128 %b
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/RISCV/half-convert.ll:9397]: Test a strictfp case?

**Comment by topperc** [comment]
> LGTM, though it's a bit confusing to see a BF16 test affected by `SoftPromoteHalfRes_FP_ROUND`.

SoftPromoteHalf was added years ago before bfloat was added. Maybe we should rename it now that it's used for bfloat too?

**Review by arsenm — APPROVED** [review]

**Pull request merged**